From 8473829d1f1f2d30d2629364204cda071749af60 Mon Sep 17 00:00:00 2001
From: Ryan Moeller <ryan@iXsystems.com>
Date: Mon, 6 Feb 2023 14:16:01 -0500
Subject: [PATCH] initramfs: Make mountpoint=none work

In initramfs, mount.zfs fails to mount a dataset with mountpoint=none,
but mount.zfs -o zfsutil works.  Use -o zfsutil when mountpoint=none.

Reviewed-by: Brian Behlendorf <behlendorf1@llnl.gov>
Reviewed-by: Richard Yao <richard.yao@alumni.stonybrook.edu>
Signed-off-by: Ryan Moeller <ryan@iXsystems.com>
Closes #14455
(cherry picked from commit eb823cbc76d28a7cafdf6a7aafdefe7e74fe26bc)
---
 contrib/initramfs/scripts/zfs | 4 +++-
 1 file changed, 3 insertions(+), 1 deletion(-)

Index: zfs/contrib/initramfs/scripts/zfs
===================================================================
--- zfs.orig/contrib/initramfs/scripts/zfs
+++ zfs/contrib/initramfs/scripts/zfs
@@ -357,9 +357,11 @@ mount_fs()
 				# isn't the root fs.
 				return 0
 			fi
-			ZFS_CMD="mount.zfs"
 			# Last hail-mary: Hope 'rootmnt' is set!
 			mountpoint=""
+			if [ "$mountpoint" = "legacy" ]; then
+				ZFS_CMD="mount.zfs"
+			fi
 		else
 			mountpoint="$mountpoint1"
 		fi
